Added info on symbolic tokens in design docs (#2657)
carbon-language/carbon-lang: docs/design/lexical_conventions/symbolic_tokens.md

| `>>`            | Arithmetic and Logical Right-shift                                                                         |
| `==`            | Equality or equal to                                                                                       |
| `!=`            | Inequality or not equal to                                                                                 |
| `>`             | Greater than                                                                                               |
| `>=`            | Greater than or equal to                                                                                   |
| `<`             | Less than                                                                                                  |
| `<=`            | Less than or equal to                                                                                      |
| `->`            | Return type and indirect member access                                                                     |
| `=>`            | Match syntax                                                                                               |

> | `[`             | Subscript and deduced parameter lists                                                                      |
> | `]`             | Subscript and deduced parameter lists                                                                      |
> | `(`             | Function call, function declaration and tuple literals                                                     |
> | `)`             | Function call, function declaration and tuple literals                                                     |
> | `{`             | Struct literals, blocks of control flow statements and the bodies of definitions (classes, functions, etc) |
> | `}`             | Struct literals, blocks of control flow statements and the bodies of definitions (classes, functions, etc) |

Reviewer: This is the kind of thing I was suggesting in my [previous comment](https://github.com/carbon-language/carbon-lang/pull/2657#discussion_r1156333901):

```suggestion
| `[` and `]`     | Subscript and deduced parameter lists                                                                      |
| `(` and `)`     | Function call, function declaration and tuple literals                                                     |
| `{` and `}`     | Struct literals, blocks of control flow statements and the bodies of definitions (classes, functions, etc) |
```

Author: Ok, I will add these!

# Symbolic Tokens

<!--
Part of the Carbon Language project, under the Apache License v2.0 with LLVM
Exceptions. See /LICENSE for license information.
SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
-->

<!-- toc -->

## Table of contents

-   [Overview](#overview)
-   [Details](#details)
    -   [Symbolic token list](#symbolic-token-list)
    -   [Whitespace](#whitespace)
-   [Alternatives considered](#alternatives-considered)
-   [References](#references)

<!-- tocstop -->

## Overview

> Symbolic tokens are a set of tokens used to represent operators. Operators are

Reviewer: As the next sentence points out, this statement isn't really accurate. How about something like this instead?

```
A _symbolic token_ is one of a fixed set of tokens that consist of characters that are not valid in identifiers.
```

> Symbolic tokens are a set of tokens used to represent operators. Operators are

Reviewer 2: ```suggestion
Symbolic tokens are tokens consisting of symbols, not letters or numbers. Operators are
```

Reviewer 2: You could also include a link to the Wikipedia definition of lexical token: https://en.wikipedia.org/wiki/Lexical_analysis#Token

Author: @josh11b @geoffromer I was thinking of the following line, and have incorporated both of your suggestion 

`A _symbolic token_ is one of a fixed set of [tokens](https://en.wikipedia.org/wiki/Lexical_analysis#Token) that consist of characters that are not valid in identifiers, that is they are tokens consisting of symbols, not letters or numbers.`

Author: @josh11b I have updated this part earlier itself! Pls check.

> Symbolic tokens are a set of tokens used to represent operators. Operators are

Reviewer 3: Symbolic token is a single character or a sequence of characters used to represent a complex or frequently used symbol.

> Symbolic tokens are a set of tokens used to represent operators. Operators are

Reviewer 3: ```suggestion
Symbolic token is a single character or a sequence of characters used to represent a complex or frequently used symbol.
```

Author: They are single characters but I think they don't necessarily represent complex or frequently used symbols.

one use of symbolic tokens, but they are also used in patterns `:`, declarations
(`->` to indicate return type, `,` to separate parameters), statements (`;`,
`=`, and so on), and other places (`,` to separate function call arguments).

Some languages have a fixed set of symbolic tokens, For example:
[C++ operators](https://eel.is/c++draft/lex.operators) and

> [rust operators](https://doc.rust-lang.org/book/appendix-02-operators.html).

Reviewer 2: ```suggestion
[Rust operators](https://doc.rust-lang.org/book/appendix-02-operators.html).
```

> While some others have extensible rules for defining operators, including the

Reviewer 2: ```suggestion
While some others have extensible rules for defining operators, leading to an
open set of symbolic tokens, including the
```

facility for a developer to define operators that aren't part of the

> baselanguage. For example:

Reviewer 2: ```suggestion
base language. For example:
```

[Swift operator rules](https://docs.swift.org/swift-book/ReferenceManual/LexicalStructure.html#ID418),

> [Haskell operator rules](https://www.haskell.org/onlinereport/haskell2010/haskellch2.html#dx7-18008).

Reviewer: I don't think this paragraph should go in the "Overview" section, because the overview should be a pretty concise statement of the key language rules that will be explained in the rest of the doc.

More broadly, I'm not sure we need this paragraph at all. The files under `docs/design` should be focused on explaining the rules of the Carbon language, and I don't think these observations about other languages really help with that.

Author: I will remove it entirely!

Carbon has a fixed set of tokens that represent operators, defined by the
language specification. Developers cannot define new tokens to represent new

> operators.

Reviewer: It seems like this paragraph should be talking about symbolic tokens, rather than tokens that represent operators.

Author: How about changing it into something like this?
`Carbon has a fixed set of symbolic tokens, defined by the language specification and is used to represent operators. An option to define a new symbolic token doesn't exist.`

Reviewer: I think we don't need "and is used to represent operators". I'd also rephrase the second sentence to be clearer that there's no way for _user code_ to define new symbolic tokens. Otherwise, it might be taken as saying that even the language designers can't add new symbolic tokens.

Author: `Developers cannot define new symbolic tokens in their own code`
or `An option to create user-defined tokens dosen't exists`

@geoffromer Will any of the above work?

Reviewer: Sure, "Developers cannot define new symbolic tokens in their own code" sounds good to me.

Author: Will add it

Symbolic tokens are lexed using a "max munch" rule: at each lexing step, the
longest symbolic token defined by the language specification that appears
starting at the current input position is lexed, if any.

Not all uses of symbolic tokens within the Carbon grammar will be treated as
operators. For example, `(` and `)` tokens serves to delimit various grammar
productions, and we may not want to consider `.` to be an operator, because its

> right "operand" is not an expression.

Reviewer: The first paragraph already explained that symbolic tokens are used for other things besides operators, so this paragraph seems unnecessary.

The presence or absence of whitespace around the symbolic token is used to
determine its fixity, in the same way we expect a human reader to recognize
them. For example, we want `a* - 4` to treat the `*` as a unary operator and the

> `-` as a binary operator, while `a * -4` treats `*` as a mathematical operation

Reviewer 2: ```suggestion
`-` as a binary operator, while `a * -4` treats `*` as a binary operation
```

> and `-` as the negative sign. Hence we can say that the whitespaces plays a

Reviewer: ```suggestion
and `-` as the negative sign. Hence we can say that the whitespace plays a
```

> and `-` as the negative sign. Hence we can say that the whitespaces plays a

Reviewer 2: ```suggestion
and `-` as the unary negative sign. Hence we can say that the whitespaces plays a
```

> really important role here, and we use some rules to avoid confusion:

Reviewer: I think we can drop this paragraph from the overview, or maybe shorten it to a single sentence like "When a symbolic token is used as an operator, the surrounding whitespace must follow certain rules:". This paragraph could work well in the "Whitespace" section below, though.

Author: So should I shorten the first paragraph to the line you mentioned and then keep the points here and then take the paragraph to the whitespace section? Or take the paragraph and point to the whitespace section, and we will mention a line saying "When a symbolic token is used as an operator, the surrounding whitespace must follow certain rules which can be found in the whitespace section" Which one would be good?

Reviewer: I'm not quite sure. Maybe let's try it with the single sentence and the bulleted list here, and see how it looks together with the other changes in this section.

-   There can be no whitespace between a unary operator and its operand.
-   The whitespace around a binary operator must be consistent: either there is
    whitespace on both sides or on neither side.
-   If there is whitespace on neither side of a binary operator, the token
    before the operator must be an identifier, a literal, or any kind of closing
    bracket (for example, `)`, `]`, or `}`), and the token after the operator
    must be an identifier, a literal, or any kind of opening bracket (for
    example, `(`, `[`, or `{`).

## Details

Symbolic tokens are intended to be used for widely-recognized operators, such as
the mathematical operators `+`, `*`, `<`, and so on. Those used as operators
would generally be expected to also be meaningful for some user-defined types,
and should be candidates for being made overloadable once we support operator

> overloading.

Reviewer: This paragraph seems like it's more about operators than about symbolic tokens (and it makes it sound like symbolic tokens are only used for operators). I'm not sure this doc needs it.

Author: josh mentioned that overloading is also supported in carbon, will it be enough if I add that too, if I add some more details will it be enough to keep the paragraph? What are your opinions @geoffromer @josh11b

Reviewer: As I understand it, this doc is supposed to be about the _syntax_  of symbolic tokens (in other words the rules that affect how they're parsed). Discussion of their _semantics_ (in other words, the _meaning_ that the language gives to them) belongs in other documents. For example, [`docs/design/expressions/arithmetic.md`](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/expressions/arithmetic.md) documents the semantics of arithmetic operators, including how to overload them.

Author: I have dropped it entirely!

> overloading.

Reviewer 2: When #601 was written, Carbon did not have any support for operator overloading. It does now, so this is probably not what we should write.
```suggestion
would generally be expected to also be meaningful for some user-defined types,
and should be overloadable.
```

Author: I also think removing the entire thing would be better as stated in the above conversation! So should I add anything else?

Reviewer 2: I'd be fine with removing this, to focus on lexical tokens rather than operators

Author: I have edited it, is it fine or should I remove that line too?

Reviewer 2: The sentence that is there now is not very relevant, and would be better removed.

Author: Done

### Symbolic token list

The following is the initial list of symbolic tokens recognized in a Carbon

> source file:

Reviewer 2: FYI, this matches what is in #601 , but a lot more symbolic tokens are defined for Carbon. A much more current list may be found in the explorer's lexer:
https://github.com/carbon-language/carbon-lang/blob/trunk/explorer/syntax/lexer.lpp

That list seems to be the most accurate. Looking at the toolchain's lexer definition: https://github.com/carbon-language/carbon-lang/blob/trunk/toolchain/lexer/token_kind.def it seems to have some things that are not officially tokens in Carbon (`<=>`, `:=`, `~=`, `<>`, `<-`, `@`, `\\`, `!`, `~`).

Author: I should add the official ones only right?

Reviewer 2: Yes. I think the explorer list is more correct than the toolchain's

| Token | Explanation                                                                                                |
| ----- | ---------------------------------------------------------------------------------------------------------- |
| `*`   | Indirection, multiplication, and forming pointers                                                          |
| `&`   | Address-of or Bitwise AND                                                                                  |
| `=`   | Assignment                                                                                                 |

> | `->`  | Return type and `p->x` equivalent to `(*p).x` (in C++)                                                     |

Reviewer: ```suggestion
| `->`  | Return type and indirect member access                                                    |
```

| `=>`  | Match syntax                                                                                               |

> | `[]`  | Subscript                                                                                                  |

Reviewer: This makes it sound like `[]` is a single token. It should probably say something like "`[` and `]`" instead. Same goes for the next two table entries.

Author: For the time being, I have given the same explanation for `[` and `]`, and for the others is it fine!?

Reviewer: I'd prefer a single table row that lists both delimiters, because it's more compact and less redundant.

> | `[]`  | Subscript                                                                                                  |

Reviewer: `[]` is also used for deduced parameter lists.

| `()`  | Function call and function declaration                                                                     |
| `{}`  | Struct literals, blocks of control flow statements and the bodies of definitions (classes, functions, etc) |

> | `,`   | Separate arguments in a function call, elements of a tuple, or parameters of a function declaration        |

Reviewer: Also fields of a struct.  As with `()`, we can simplify this to something like "Separate tuple and struct elements", because function argument lists are just tuples.

| `.`   | Member access                                                                                              |

> | `:`   | Scope                                                                                                      |

Reviewer 2: ```suggestion
| `:`   | Bindings                                                                                                   |
```

> | `:`   | Scope                                                                                                      |

Reviewer: "Scope" doesn't seem accurate. I'd say something like "Name binding" instead.

> 

Reviewer: `;` seems to be missing from this list.

This list is expected to grow over time as more symbolic tokens are required by
language proposals.

> Note: The above list only covers up to

Reviewer: This should probably be a TODO rather than a note. Or, if you want to fix it right away, the missing tokens I noticed are:
- [Arithmetic operators](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/expressions/arithmetic.md)
- [Bitwise operators](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/expressions/bitwise.md)
- [Comparison operators](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/expressions/comparison_operators.md)
- [`:!`](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/README.md#checked-and-template-parameters)

Author: I will be adding all the tokens present here https://github.com/carbon-language/carbon-lang/blob/trunk/explorer/syntax/lexer.lpp

It will cover all these right?

Reviewer: I'm not sure. The explorer might not support everything in the design yet, and sometimes for various reasons we extend the Explorer with things that aren't in the design yet. I think we should use the other design docs as our source of truth, not the Explorer code.

[#601](https://github.com/carbon-language/carbon-lang/pull/601) and more have
been added since that are not reflected here.

> ### Whitespace

Reviewer: This whole section seems to be focused on explaining the background and rationale for this part of the design, but it should be focused primarily on what the design _is_. In other words, the focus should be on what a reader needs to know in order to read and write Carbon code. See, for example, the discussion of [real number literals](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/lexical_conventions/numeric_literals.md#real-number-literals): it starts by explaining the syntax and semantics of real-number literals, and then at the end of the section it goes back to explain the rationale for one particularly noteworthy design choice (namely, the decision to support hexadecimal literals).

Author: @geoffromer I have revamped this section! Can you take a look at it!?

Reviewer: It still seems kind of focused on the rationale, especially at the beginning and the end. But after thinking about it more, I'm not sure we need this section at all: the overview already explains the whitespace rules quite clearly, so I'm not sure there's much more we need to say. It would be good to at least mention the rationale behind these rules, but I think we can do it with a sentence at the end of the overview, such as "These rules enable us to use a token like `*` as a prefix, infix, and postfix operator, without creating ambiguity."

Author: Will make the changes

> to support the use of the same symbolic token as a prefix operator, an infix

Reviewer: ```suggestion
To support the use of the same symbolic token as a prefix operator, an infix
```

to simply and unambiguously parse operators that might have all three fixities.

For example, given the expression `a * - b`, there are two possible parses:

-   As `a * (- b)`, multiplying `a` by the negation of `b`.
-   As `(a *) - b`, subtracting `b` from the pointer type `a *`.

The chosen rule to distinguish such cases is to consider the presence or absence
of whitespace, as it strikes a good balance between simplicity and
expressiveness for the programmer and simplicity and good support for error
recovery in the implementation. Hence `a * -b` uses the first interpretation,
`a* - b` uses the second interpretation, and other combinations (`a*-b`,
`a *- b`, `a* -b`, `a * - b`, `a*- b`, `a *-b`) are rejected as errors.

We require whitespace to be present or absent around the operator to indicate
its fixity, as this is a cue that a human reader would use to understand the
code: binary operators have whitespace on both sides, and unary operators lack
whitespace between the operator and its operand.

But in some cases omitting the whitespace around a binary operator aids
readability, such as in expressions like `2*x*x + 3*x + 1`, hence we have an

> allowance in such cases. In this case the operator with whitespace on neither

Reviewer 2: ```suggestion
allowance for such cases. In this case the operator with whitespace on neither
```

> But in some cases omitting the whitespace around a binary operator aids
> readability, such as in expressions like `2*x*x + 3*x + 1`, hence we have an
> allowance in such cases. In this case the operator with whitespace on neither
> side, if the token immediately before the operator indicates it is the end of an
> operand, and the token immediately after the operator indicates it is the
> beginning of an operand, the operator is treated as binary.

Reviewer 3: ```suggestion
While it is generally recommended to use whitespace to improve readability, there are cases where omitting whitespace can actually aid in readability. One such case is when you have a long expression with nested parentheses and operators.
`(2*x+1)/((x+1)*(x-1))-3*(x-2)`

While this expression is valid and follows the rules of the Carbon syntax, it can be difficult to read due to the nested parentheses and operators. However, if we omit whitespace around the binary operators, we can make the expression more readable:
`(2*x + 1) / ((x + 1) * (x - 1)) - 3 * (x - 2)`

In this version of the expression, the lack of whitespace around the operators actually makes it easier to see the grouping of terms, and the overall structure of the expression.
```

Reviewer: This document is part of the language reference, so it needs to focus on what the rules of the language are, and why, rather than on things like style guidance.

Reviewer 2: I disagree -- this is not a language reference but a design document, and rationale is definitely part of that.

Reviewer: Yes, that's why I included "and why", but by my reading @satakshigarg's suggestion is not giving a rationale, it's giving advice about how best to use this feature.

Reviewer: Just to be clear, I'm saying that I don't think any change is needed here.

<!-- omit in toc -->

# Symbolic Tokens

<!--
Part of the Carbon Language project, under the Apache License v2.0 with LLVM
Exceptions. See /LICENSE for license information.
SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
-->

<!-- toc -->
<!-- omit in toc -->

## Table of contents

-   Overview
-   Details
-   Alternatives considered
-   References

<!-- tocstop -->

## Overview

Symbolic tokens are a set of tokens used to represent operators. Operators are

> one use of symbolic tokens, but they are also used in patterns (:), declarations
> (-> to indicate return type, to separate parameters), statements (;, =, and so
> on), and other places (, to separate function call arguments).

Reviewer 4: ```suggestion
one use of symbolic tokens, but they are also used in patterns `:`, declarations
(`->` to indicate return type, `,` to separate parameters), statements (`;`, `=`, and so
on), and other places (`,` to separate function call arguments).
```

Author: Yeah! I forgot to add these! Thx!

and should be candidates for being made overloadable once we support operator
overloading.

### Symbolic token list

The following is the initial list of symbolic tokens recognized in a Carbon
source file:

| Token | Explanation                                                                                                |
| ----- | ---------------------------------------------------------------------------------------------------------- |
| `*`   | Indirection, multiplication, and forming pointers                                                          |
| `&`   | Address-of or Bitwise AND                                                                                  |
| `=`   | Assignment                                                                                                 |
| `->`  | Return type and `p->x` equivalent to `(*p).x` (in C++)                                                     |
| `=>`  | Match syntax                                                                                               |
| `[]`  | Subscript                                                                                                  |

> | `()`  | Function call and function declaration                                                                     |

Reviewer 4: what about tuple formation?

> | `()`  | Function call and function declaration                                                                     |

Reviewer 4: ```suggestion
| `()`  | Function call and function declaration and tuple literals                                                 |
```

Author: This was taken from [here](https://github.com/carbon-language/carbon-lang/pull/2519/files#r1086858109) and it was corrected by @josh11b so, not sure whether I should add it!

Reviewer: You should add it. Josh was just saying to add "function declaration"; he wasn't saying to not add anything else. Speaking of which, this list should also have "expression grouping", as in `(x + y) * z`.

We could probably simplify the list to say "Expression grouping and tuple formation", because the `()`s in function calls and function declarations are actually just tuple literals (or tuple patterns).

Alternatively, maybe we should avoid this whole issue and just have a list of operators, without explaining their semantics, like in `p0601.md`?

Reviewer 2: #2519 started by trying to document "operator tokens", but we eventually decided that wasn't the way to go. For example, `->` is sometimes an operator (in `x->y`) and sometimes not (in `fn F() -> bool`). Instead, we decided that the most useful thing was to consider this from a lexing perspective, which does not distinguish whether the token is an operator, except whether it has whitespace on each side. As long as the presence or absence of whitespace information is preserved, lexing can treat the tokens uniformly. See the discussion here, which is very relevant to this PR:
https://discord.com/channels/655572317891461132/709488742942900284/1067599248813740072

With this shift to talking about symbolic tokens, it is fair to say `()` is also used for tuple literals.

Author: Really sorry for the late reply, got held up in urgent work! I will add these!

## Overview

A _symbolic token_ is one of a fixed set of
[tokens](https://en.wikipedia.org/wiki/Lexical_analysis#Token) that consist of

> characters that are not valid in identifiers, that is they are tokens consisting

Reviewer: ```suggestion
characters that are not valid in identifiers. That is, they are tokens consisting
```

The following is the initial list of symbolic tokens recognized in a Carbon
source file:

| Symbolic Tokens and Explanation                                                                                |
| -------------------------------------------------------------------------------------------------------------- | ------------ |

> | `+` Addition                                                                                                   |

Reviewer: It looks like the `|` separators in this table are missing. This might be related to my comment below on "Bitwise OR", so probably fix that first.

| `-` Subtraction and negation                                                                                   |

> | `*` Indirection, multiplication, and forming pointers                                                          |

Reviewer: ```suggestion
| `*` Indirection, multiplication, and forming pointer types                                                          |
```

| `/` Division                                                                                                   |
| `%` Modulus                                                                                                    |
| `=` Assignment                                                                                                 |
| `^` Complementing and Bitwise XOR                                                                              |
| `&` Address-of and Bitwise AND                                                                                 |

> | `                                                                                                              | ` Bitwise OR |

Reviewer: It looks like this needs to be written with a backslash, like this:

```markdown
| `\|` | Bitwise OR |
```

Otherwise, GitHub (and the auto-formatter) will treat the `|` as a table separator.

Author: Yeah! I forgot that! didnt notice it!

Author: @geoffromer I removed the entire [section](https://github.com/carbon-language/carbon-lang/pull/2657#discussion_r1182973856) based on an earlier discussion! Any changes required?

Reviewer: > @geoffromer I removed the entire [section](https://github.com/carbon-language/carbon-lang/pull/2657#discussion_r1182973856) based on an earlier discussion! Any changes required?

Did you mean to reply to a different comment? This comment is about a section that hasn't been removed, but it looks like it's been resolved, so no more change is needed for this comment. There are some other comments that haven't been addressed yet, though (in particular, the table is still missing the `|` separators).

Author: @geoffromer I didn't understand [this](https://github.com/carbon-language/carbon-lang/pull/2657#discussion_r1156333901) comment properly, thats why I created a table like that with one column. I have updated it, is this right?

| `[` Subscript and deduced parameter lists                                                                      |
| `]` Subscript and deduced parameter lists                                                                      |
| `(` Function call, function declaration and tuple literals                                                     |
| `)` Function call, function declaration and tuple literals                                                     |
| `{` Struct literals, blocks of control flow statements and the bodies of definitions (classes, functions, etc) |
| `}` Struct literals, blocks of control flow statements and the bodies of definitions (classes, functions, etc) |
| `,` Separate tuple and struct elements                                                                         |
| `.` Member access                                                                                              |
| `:` Name bindings                                                                                              |
| `;` Statement separator                                                                                        |

> | `:!` Type-checking                                                                                             |

Reviewer: I think "Generic binding" would be more accurate. This should probably go next to `:`, since the two are very closely related.

The following is the initial list of symbolic tokens recognized in a Carbon
source file:

> | Symbolic Tokens | Explanation                                                                                                |

Reviewer 2: Doesn't look like most of the assignment operators from proposal #2511 have been included. Is that intentional?

Author: @josh11b Actually I added the operators according to [this](https://github.com/carbon-language/carbon-lang/pull/2657#discussion_r1137826711). So shall I add the ones in #2511 too???

Reviewer 2: Or add a note that the ones in #2511 are still to be added.

Author: @josh11b I have added the message.

| `==`            | Equality or equal to                                                                                       |
| `!=`            | Inequality or not equal to                                                                                 |
| `>`             | Greater than                                                                                               |
| `>=`            | Greater than or equal to                                                                                   |
| `<`             | Less than                                                                                                  |
| `<=`            | Less than or equal to                                                                                      |
| `->`            | Return type and indirect member access                                                                     |
| `=>`            | Match syntax                                                                                               |
| `[` and `]`     | Subscript and deduced parameter lists                                                                      |

> | `(` and `)`     | Function call, function declaration and tuple literals                                                     |

Reviewer 2: ```suggestion
| `(` and `)`     | Function call, function declaration, and tuple literals                                                     |
```

> | `{` and `}`     | Struct literals, blocks of control flow statements and the bodies of definitions (classes, functions, etc) |

Reviewer 2: ```suggestion
| `{` and `}`     | Struct literals, blocks of control flow statements, and the bodies of definitions (classes, functions, etc.) |
```

| `*`   | Indirection, multiplication, and forming pointers                                                          |
| `&`   | Address-of or Bitwise AND                                                                                  |
| `=`   | Assignment                                                                                                 |
| `->`  | Return type and indirect member access                                                                     |
| `=>`  | Match syntax                                                                                               |
| `[`   | Subscript and used for deduced parameter lists                                                             |

> | `]`   | Subscript and used for deduced parameter lists                                                             |

Reviewer: ```suggestion
| `[`   | Subscript and deduced parameter lists                                                             |
| `]`   | Subscript and deduced parameter lists                                                             |
```

| `(`   | Separate tuple and struct elements                                                                         |

> | `)`   | Separate tuple and struct elements                                                                         |

Reviewer: The description seems like it was copied from `,` by accident.

Author: Oh yeah! It happened accidentally!

| `{`   | Struct literals, blocks of control flow statements and the bodies of definitions (classes, functions, etc) |
| `}`   | Struct literals, blocks of control flow statements and the bodies of definitions (classes, functions, etc) |
| `,`   | Separate tuple and struct elements                                                                         |
| `.`   | Member access                                                                                              |
| `:`   | Name bindings                                                                                              |

> | `;`   | Name bindings                                                                                              |

Reviewer: ```suggestion
| `;`   | Statement separator                                                                                              |
```
